### PR TITLE
🎉 Source Hubspot: update user-defined custom field schema generation

### LINKED ISSUE #4779: Error Syncing between Hubspot to BigQuery
## Enviroment
- **Airbyte version**:  v0.27.2-alpha
- **OS Version / Instance**: GCP
- **Deployment**: GKE
- **Source Connector and version**: Hubspot 0.1.5
- **Destination Connector and version**: BigQuery 0.3.7
- **Severity**: Medium
- **Step where error happened**: Sync Job

## Current Behavior
Airbyte syncs the source, but then fails to parse some values for bigquery gets a lot of these errors
Bad double value: true
Looks very similiar to this https://github.com/airbytehq/airbyte/issues/1904

## Expected Behavior
Sync should work fine, as it works fine with bigquery denormalised destination, so issue must be in the sql

### PR DESCRIPTION
**Pull request by gaart**
## What
Change the custom fields handler to produce valid JSON schema definitions.
Fix for #4779

## How
Validate incoming types, fix the ones we know, fallback for the unknown with a warning.

## Recommended reading order
1. `x.python`

## Pre-merge Checklist
Expand the checklist which is relevant for this PR. 

<details><summary> <strong> Connector checklist </strong> </summary>
<p>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Secrets are annotated with `airbyte_secret` in the connector's spec
- [ ] Credentials added to Github CI if needed and not already present. [instructions for injecting secrets into CI](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#using-credentials-in-ci). 
- [ ] Unit & integration tests added as appropriate (and are passing)
    * Community members: please provide proof of this succeeding locally e.g: screenshot or copy-paste acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] `/test connector=connectors/<name>` command as documented [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-an-existing-connector) is passing. 
    * Community members can skip this, Airbyters will run this for you. 
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] `README.md`
    - [ ] `docs/SUMMARY.md` if it's a new connector
    - [ ] Created or updated reference docs in `docs/integrations/<source or destination>/<name>`.
    - [ ] Changelog in the appropriate page in `docs/integrations/...`. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md` contains a reference to the new connector
    - [ ] Build status added to [build page](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/builds.md)
- [ ] Build is successful
- [ ] Connector version bumped like described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [ ] No major blockers
- [ ] PR merged into master branch
- [ ] Follow up tickets have been created
- [ ] Associated tickets have been closed & stakeholders notified
</p>
</details>

<details><summary> <strong> Connector Generator checklist </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py [hunk 1]
```diff
@@ -303,8 +315,11 @@ def properties(self) -> Mapping[str, Any]:
         data = self._api.get(f"/properties/v2/{self.entity}/properties")
         for row in data:
             field_type = row["type"]
-            if field_type in ["enumeration", "date", "date-time"]:
-                field_type = "string"
+            if field_type not in VALID_JSON_SCHEMA_TYPES:
+                field_type = KNOWN_CONVERTIBLE_SCHEMA_TYPES.get(field_type) or None
+                if not field_type:
+                    logger.warn(f"Unsupported type {field_type} converted to string")
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:321]: If the `field_type` is not recognized the variable will be `None`, maybe something:

```python
if not KNOWN_CONVERTIBLE_SCHEMA_TYPES.get(field_type):
    log.warn("not recognized type_field will be converted to string") 

field_type = KNOWN_CONVERTIBLE_SCHEMA_TYPES.get(field_type) or "string"
```

**Review by marcosmarxm — DISMISSED** [review]
Thanks for the contribution!

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py [hunk 2]
```diff
@@ -303,8 +315,11 @@ def properties(self) -> Mapping[str, Any]:
         data = self._api.get(f"/properties/v2/{self.entity}/properties")
         for row in data:
             field_type = row["type"]
-            if field_type in ["enumeration", "date", "date-time"]:
-                field_type = "string"
+            if field_type not in VALID_JSON_SCHEMA_TYPES:
+                field_type = KNOWN_CONVERTIBLE_SCHEMA_TYPES.get(field_type) or None
+                if not field_type:
```
Comment (keu) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:320]: what about `format` attribute?
I think we should have a copy of original type there. WDYT?

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py [hunk 3]
```diff
@@ -303,8 +315,11 @@ def properties(self) -> Mapping[str, Any]:
...
+                    logger.warn(f"Unsupported type {field_type} converted to string")
+                    field_type = "string"
```
Comment (keu) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:322]: Also, this will lead to failure of the destination if actual type in record != `string`.
The possible solution can be a lookup during parsing response and convert type accordingly to string if its type is string. Also probably `property` needs to be replaced with `cached_property`

**Review by keu — CHANGES_REQUESTED** [review]
small comments

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py [hunk 4]
```diff
@@ -40,6 +40,18 @@
 # we got this when provided API Token has incorrect format
 CLOUDFLARE_ORIGIN_DNS_ERROR = 530
 
+VALID_JSON_SCHEMA_TYPES = {"string", "integer", "number", "boolean", "object", "array", "bit"}
```
Comment (avida) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:43]: What is bit type? Cant find it definition for json-schema https://json-schema.org/understanding-json-schema/reference/type.html

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py [hunk 5]
```diff
@@ -40,6 +40,18 @@
...
+
+KNOWN_CONVERTIBLE_SCHEMA_TYPES = {
+    "bool": "boolean",
+    "enumeration": "string",
+    "date": "string",
+    "date-time": "string",
+    "datetime": "string",
+    "json": "object",
```
Comment (avida) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:51]: I don't think its a good idea to treat unknown custom data as an object. 
a) It could be a string containing invalid json
b) If this an object then json schema is expected to enumerate  list of properties 
c) In case is destination is relational DB with normalization turned on it would create unknown amount of tables for nested structure. 

I think its better to treat it as a string.

Comment (keu) [reply airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:51]: good catch

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py (hunk 4 above)
Comment (gaart) [reply airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:43]: It's supported by airbyte https://docs.airbyte.io/understanding-airbyte/basic-normalization
Anyway, I'll remove it for a better interoperability. Thanks

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py (hunk 5 above)
Comment (gaart) [reply airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:51]: Agree

**Review by keu — APPROVED** [review]

File: airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
+import pytest
+from source_hubspot.api import Stream
+
+
+@pytest.mark.parametrize("field_type,expected", [
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py:5]: 💪🏼 tests ❤️

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py [hunk 7]
```diff
@@ -293,7 +312,30 @@ def read(self, getter: Callable, params: Mapping[str, Any] = None) -> Iterator:
...
+    def _get_field_props(field_type: str) -> Mapping[str, str]:
+
+        if field_type in VALID_JSON_SCHEMA_TYPES:
+            return {
+                "type": field_type,
+            }
+
+        converted_type, field_format = KNOWN_CONVERTIBLE_SCHEMA_TYPES.get(field_type) or (None, None)
+
+        if not converted_type:
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:325]: let's default to string if type is unknown to avoid failures
```suggestion
        if not converted_type:
            converted_type = "string"
```

**Review by sherifnada — APPROVED** [review]
@gaart looks great! Could you make sure to follow the "Connector Checklist" under the "pre-merge checklist" to add the following items: 

1. 🎉 to the PR title
2. changelog to docs

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py (hunk 7 above)
Comment (gaart) [reply airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:325]: @sherifnada in case we return `converted_type = "string"` in the schema, we should validate and convert the data at the transform step for each record, don't we? Meaning it can be an irreversible transformation for those fields which requires extra steps with no guarantee it works for every case. 

Can we just return `converted_type = "string"` for unknown fields without actual validation and converting for them?

@keu what do you think?

Comment (sherifnada) [reply airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:325]: @gaart my understanding is that without doing something like the suggestion above, the output schema will contain a field without any `type` field because of the `(None, None)` statement above. 

If we need to cast to a string that's acceptable IMO. alternatively we can skip the field altogether

Comment (keu) [reply airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:325]: casting to string seems safe enough for unknown data types, which should be rare, we probably want some warning as well

File: airbyte-integrations/connectors/source-hubspot/acceptance-test-config.yml [hunk 8]
```diff
@@ -5,8 +5,8 @@ tests:
   connection:
     - config_path: "secrets/config.json"
       status: "succeed"
-    - config_path: "integration_tests/invalid_config.json"
-      status: "exception"
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-hubspot/acceptance-test-config.yml:9]: why was this removed? we should put it back

**Review by sherifnada — APPROVED** [review]

File: airbyte-integrations/connectors/source-hubspot/acceptance-test-config.yml (hunk 8 above)
Comment (gaart) [reply airbyte-integrations/connectors/source-hubspot/acceptance-test-config.yml:9]: It's failing, and the best way to fix it I think is to migrate this connector to the CDK, here's the issue https://github.com/airbytehq/airbyte/issues/5114

Comment (sherifnada) [reply airbyte-integrations/connectors/source-hubspot/acceptance-test-config.yml:9]: why is it failing?

Comment (gaart) [reply airbyte-integrations/connectors/source-hubspot/acceptance-test-config.yml:9]:
```

self = <source_acceptance_test.tests.test_core.TestConnection object at 0x7fefd1176b90>
connector_config = SecretDict(******)
inputs = ConnectionTestConfig(config_path='integration_tests/invalid_config.json', status=<Status.Exception: 'exception'>, timeout_seconds=None)
docker_runner = <source_acceptance_test.utils.connector_runner.ConnectorRunner object at 0x7fefd1176210>

    def test_check(self, connector_config, inputs: ConnectionTestConfig, docker_runner: ConnectorRunner):
 
    ...
    
            assert err.value.exit_status != 0, "Connector should exit with error code"
>           assert "Traceback" in err.value.stderr.decode("utf-8"), "Connector should print exception"
E           AttributeError: 'str' object has no attribute 'decode'

/usr/local/lib/python3.7/site-packages/source_acceptance_test/tests/test_core.py:92: AttributeError
=========================== short test summary info ============================
FAILED test_core.py::TestConnection::test_check[inputs1] - AttributeError: 's...
============== 1 failed, 8 passed, 2 skipped in 291.69s (0:04:51) ==============
```

Comment (sherifnada) [reply airbyte-integrations/connectors/source-hubspot/acceptance-test-config.yml:9]: should this be just "failed" then? See https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py#L74

Comment (gaart) [reply airbyte-integrations/connectors/source-hubspot/acceptance-test-config.yml:9]: You're right, changed, works now, thank you!
